packages/flutter/lib/src/rendering/editable_line.dart

```diff
@@ -284,17 +293,26 @@ class RenderEditableLine extends RenderBox {
   @override
   void performLayout() {
     Size oldSize = hasSize ? size : null;
-    double lineHeight = constraints.constrainHeight(_preferredHeight);
+    double lineHeight = constraints.constrainHeight(_preferredLineHeight);
     _caretPrototype = new Rect.fromLTWH(0.0, _kCaretHeightOffset, _kCaretWidth, lineHeight - 2.0 * _kCaretHeightOffset);
     _selectionRects = null;
     _textPainter.layout(maxWidth: _maxContentWidth);
-    size = new Size(constraints.maxWidth, constraints.constrainHeight(math.max(lineHeight, _textPainter.height)));
+    size = new Size(constraints.maxWidth, constraints.constrainHeight(
+      _textPainter.height.clamp(lineHeight, lineHeight * _maxLines)));
     Size contentSize = new Size(_textPainter.width + _kCaretGap + _kCaretWidth, _textPainter.height);
-    if (onPaintOffsetUpdateNeeded != null && (size != oldSize || contentSize != _contentSize))
-      onPaintOffsetUpdateNeeded(new ViewportDimensions(containerSize: size, contentSize: contentSize));
+    assert(_selection != null);  // valid assumption?
```

Comment:
(one option is to just say that it is, and treat it as a bug if we find it isn't later)